codegen/IRTS/CodegenVim.hs

```diff
@@ -23,17 +23,66 @@ import qualified Vimscript.Render          as Vim
 
 codegenVim :: CodeGenerator
 codegenVim ci = do
-  let prg = runReader (genProgram (simpleDecls ci)) HM.empty
+  let decls = simpleDecls ci
+  prg <- runReaderT (genProgram decls) HM.empty
   writeFile (outputFile ci) (pretty 200 (Vim.renderProgram prg))
 
-type Gen a = Reader (HashMap Vim.Name Vim.ScopedName) a
+type Gen a = ReaderT (HashMap Vim.Name Vim.ScopedName) IO a
 
 vimName :: Idris.Name -> Vim.Name
-vimName n = Vim.Name ("Idris_" <> foldMap vimChar (showCG n))
+vimName n = Vim.Name (T.pack ("Idris_" <> foldMap vimChar (showCG n)))
   where
-    vimChar x
-      | isAlpha x || isDigit x = T.singleton x
-      | otherwise = "_" <> T.pack (show (fromEnum x)) <> "_"
+  vimChar x
+    | isAlpha x || isDigit x = [x]
+    | otherwise = encode_digit_ch x
+
+  unencodedChar :: Char -> Bool   -- True for chars that don't need encoding
+  unencodedChar 'Z' = False
+  unencodedChar 'z' = False
+  unencodedChar c   =  c >= 'a' && c <= 'z'
+                    || c >= 'A' && c <= 'Z'
+                    || c >= '0' && c <= '9'
+
+  encode_digit_ch :: Char -> String
+  encode_digit_ch c | c >= '0' && c <= '9' = encode_as_unicode_char c
+  encode_digit_ch c | otherwise            = encode_ch c
+
+  encode_ch :: Char -> String
+  encode_ch c | unencodedChar c = [c]
+  -- encode_ch '('  = "ZL"   
+  -- encode_ch ')'  = "ZR"   
+  -- encode_ch '['  = "ZM"
+  -- encode_ch ']'  = "ZN"
+  encode_ch ':'  = "ZC"
+  encode_ch 'Z'  = "ZZ"
+  encode_ch 'z'  = "zz"
+  encode_ch '&'  = "za"
+  encode_ch '|'  = "zb"
+  encode_ch '^'  = "zc"
+  encode_ch '$'  = "zd"
+  encode_ch '='  = "ze"
+  encode_ch '>'  = "zg"
+  encode_ch '#'  = "zh"
+  encode_ch '.'  = "zi"
+  encode_ch '<'  = "zl"
+  encode_ch '-'  = "zm"
+  encode_ch '!'  = "zn"
+  encode_ch '+'  = "zp"
+  encode_ch '\'' = "zq"
+  encode_ch '\\' = "zr"
+  encode_ch '/'  = "zs"
+  encode_ch '*'  = "zt"
+  encode_ch '_'  = "zu"
+  encode_ch '%'  = "zv"
+  encode_ch c    = encode_as_unicode_char c
+
+  encode_as_unicode_char :: Char -> String
+  encode_as_unicode_char c = 'z' : if isDigit (head hex_str) then hex_str
+                                                             else '0':hex_str
+    where hex_str = showHex (ord c) "U"
+    -- ToDo: we could improve the encoding here in various ways.
+    -- eg. strings of unicode characters come out as 'z1234Uz5678U', we
+    -- could remove the 'U' in the middle (the 'z' works as a separator).
```

Comment:
Would be nice to break out the encoding stuff into a separate module, and have them named with camel casing.

Reply:
Done in https://github.com/owickstrom/idris-vimscript/pull/1/commits/bdcaf0dd6f4cdfc0d89de03bf26afd565586b5ab